specification/window.browser.bs

```diff
@@ -0,0 +1,57 @@
+<pre class="metadata">
+Title: window.browser
+Shortname: wecg-browser
+Level: 1
+Group: wecg
+Status: UD
+URL: https://patrickkettner.github.io/window.browser
+Editor: Patrick Kettner, Google, patrickkettner@google.com
+Abstract: This specification reserves the <code>window.browser</code> namespace for use by WebExtensions.
+Repository: patrickkettner/window.browser
+Complain About: accidental-2119 on, missing-example-ids on
+</pre>
+
+<pre class=link-defaults>
+spec:html; type:attribute; text:window
+</pre>
+
+<pre class="anchors">
+spec: html; urlPrefix: https://html.spec.whatwg.org/multipage/
+    type: dfn; text: valid non-empty URL; url: valid-non-empty-url-potentially-surrounded-by-spaces
+</pre>
+
+<section class="non-normative">
+  ## Introduction {#introduction}
+  <em>This section is non-normative.</em>
+
+  Modern WebExtensions are an evolution of a system originally introduced in
+  Chrome version 3. In that system, any API that was added was put under the
+  <code>chrome</code> namespace. When WebExtensions were adopted outside of
+  Chromium, other vendors used the more neutral <code>browser</code> namespace.
+  That <code>browser</code> is now considered to canonically correct namespace
+  for all WebExtension APIs.  Some interactions with WebExtensions (such as
+  external messaging from content scripts to background processes) necessitate
+  the exposure of this namespace to the web.
+
+  Given that, this document defines <code>window.browser</code> as reserved for
+  the use of APIs designed to interact with WebExtensions. The specifics of what
+  APIs within <code>window.browser</code> are at present intentionally undefined,
+  and are left up to User Agent implementation.
+</section>
+
+
+  <section>
+    <h3 id="window-interface">
+      <a attribute lt="browser"><code>window.browser</code></a> API
+    </h3>
+
+  {{browser}} is UA defined attribute exposed on {{window}}. When implemented,
+  it MUST be used only for WebExtension related functionality.
+  </section>
+
+
+  <pre class="idl">
+  partial interface Window {
+    readonly attribute object browser;
```

Comment:
```suggestion
   attribute object browser;
```

Suggestion to remove `readonly`, for the following reasons:

- These properties are currently not read-only, at least not in Chrome, nor in Firefox.
- Making it writable reduces the webcompat risk. Using `readonly` breaks websites that assign to `browser`.
- Making it writable also makes it easier to replace the `browser` namespace in case one wants to replace it altogether to support unit tests.

A counterpoint, in favor of adding `readonly`, is that it enables web pages to rely on the presence of an unmodified `browser`, unaffected by other browser extensions or third-party libraries, for example. I am not attributing much value to this feature, because one cannot expect any protections once third-party code has executed in the context.

Reply:
Although my suggested edit has been approved, I'd like to temporarily re-open this thread to allow other vendors to chime in. In particular, I'd like to hear @xeenon 's perspective on Safari's behavior. I see that Safari's interface of the `browser` namespace content itself is defined at https://github.com/WebKit/WebKit/blob/ea5e30e46342b0e755281a1ffb2f23c37a2b9bcd/Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPINamespace.idl , with widespread use of "readonly", but cannot tell from the WebKit source code whether the `browser` global is supposed to be `readonly` or writable.

Reply:
window.browser is not readonly in WebKit / Safari.

Reply:
it is also not readonly in Chrome. I would be surprised if there was interest in changing it.